services/background-worker/src/Wire/Defederation.hs

```diff
@@ -53,75 +50,65 @@ deleteFederationDomainInner' go (msg, envelope) = do
-      req :: Domain -> Request
-      req dom =
-        defaultRequest
-          { method = methodDelete,
-            secure = False,
-            host = galley env ^. epHost . to encodeUtf8,
-            port = galley env ^. epPort . to fromIntegral,
-            path = "/i/federation/" <> toByteString' dom,
-            requestHeaders = ("Accept", "application/json") : requestHeaders defaultRequest,
-            responseTimeout = defederationTimeout env
-          }
-  let callGalley d = do
-        -- Retry the request a couple of times. If the final one fails, catch the exception\
-        -- so that we can NACK the message and requeue it.
-        resp <- try $ recovering x3 httpHandlers $ \_ -> liftIO $ httpLbs (req d) manager
-        either
-          -- Requeue the exception and rethrow the exception
-          (\(e :: SomeException) -> liftIO (reject envelope True) >> throwM e)
-          go
-          resp
+deleteFederationDomainInner :: (RabbitMQEnvelope e) => MVar () -> (Q.Message, e) -> AppT IO ()
+deleteFederationDomainInner runningFlag (msg, envelope) =
   deleteFederationDomainInner' (const callGalley) (msg, envelope)
   where
-    go :: Response L.ByteString -> AppT IO ()
-    go resp = do
-      let code = statusCode $ responseStatus resp
-      if code >= 200 && code <= 299
-        then do
-          liftIO $ ack envelope
-        else -- ensure that the message is requeued
-        -- This message was able to be parsed but something
-        -- else in our stack failed and we should try again.
-          liftIO $ reject envelope True
+    callGalley d = do
+      env <- ask
+      -- Jittered exponential backoff with 10ms as starting delay and 300s as max
+      -- delay. When 300s is reached, every retry will happen after 300s.
+      let policy = capDelay 300_000_000 $ fullJitterBackoff 10000
+          manager = httpManager env
+      recovering policy httpHandlers $ \_ ->
+        bracket_ (takeMVar runningFlag) (putMVar runningFlag ()) $ do
+          -- Non 2xx responses will throw an exception
+          -- So we are relying on that to be caught by recovering
+          resp <- liftIO $ httpLbs (req env d) manager
+          let code = statusCode $ responseStatus resp
+          if code >= 200 && code <= 299
+          then do
+            liftIO $ ack envelope
+          else -- ensure that the message is requeued
+          -- This message was able to be parsed but something
+          -- else in our stack failed and we should try again.
+            liftIO $ reject envelope True
+    req env dom =
+      defaultRequest
+        { method = methodDelete,
+          secure = False,
+          host = galley env ^. epHost . to encodeUtf8,
+          port = galley env ^. epPort . to fromIntegral,
+          path = "/i/federation/" <> toByteString' dom,
+          requestHeaders = ("Accept", "application/json") : requestHeaders defaultRequest,
+          responseTimeout = defederationTimeout env
```

Comment:
Looks like a long running http request, this is likely going to blow up for a big enough instance. I think we need to reconsider this approach, but perhaps in a separate PR.